Portal lock users take 2

britneywright opened the pull request:
Based on @pauljamescleary PR I added the check of the user's "isLocked" attribute in the custom action builders. There's two action builders now, one for the API and one for the Frontend. It might be overkill to have the custom action on the api calls in VinylDNS.scala with it handled in the views and the API itself will check the user account when that PR is merged. I didn't update the tests yet with that in mind.

Overall I like the custom action builder approach. reduced a lot of redundancy in the code.

pauljamescleary commented on modules/portal/app/controllers/VinylDNS.scala:
```diff
@@ -364,193 +329,165 @@ class VinylDNS @Inject()(
+    val vinyldnsRequest =
+      new VinylDNSRequest("POST", s"$vinyldnsServiceBackend", "zones", payload)
+    val signedRequest =
+      signRequest(vinyldnsRequest, getUserCreds(request.session.get("accessKey")))
```
we shouldn't need to call `getUserCreds` any more, as the `implicit request` provided should be a `UserAccountRequest` and have the `account` already available.  Eliminates a call.

Also, there is a pattern to all of these, namely:

1. Sign request
2. Execute
3. Send response back out with cache headers.

We can change the `executeRequest` to be 

```scala
def executeRequest(req: VinylDNSRequest, user: UserAccount)
```

and do the signing in there, as well as the `.map(response => Status(response.status)...withHeaders` bits all in there.

mitruly commented on modules/portal/app/models/UserAccount.scala:
```diff
@@ -30,7 +32,8 @@ case class UserAccount(
+    isLocked: Boolean = false) {
 
   private def generateKey: String = RandomStringUtils.randomAlphanumeric(20)
 
```
There's an unused import on line 90.

mitruly commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,99 @@
+ */
+
+package controllers
+import controllers.VinylDNS.Alerts
```
Both of these imports are unused:
```
import controllers.VinylDNS.Alerts
import javax.inject.Inject
```

mitruly commented:
I think I'm good on this once the build passes.

britneywright commented on modules/portal/app/controllers/VinylDNS.scala:
```diff
@@ -233,32 +211,24 @@ class VinylDNS @Inject()(
-        case Failure(ex) => throw ex
-      }
-    }
+  def serveCredsFile(fileName: String): Action[AnyContent] = frontendAction.async {
```
I had to use the frontendAction here otherwise the portal renders the csv path as a blank screen with just the text "Account is Locked", now it redirects to the login screen if a user is locked and tries to download credentials.

pauljamescleary replied:
answered my question, thanks.

pauljamescleary commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,98 @@
+            }
+        }
+    }
+}
```
there is duplication here that could be refactored.  I am ok with leaving it alone for the time being as it is unclear how the two actions may evolve.

pauljamescleary commented on modules/portal/app/controllers/VinylDNS.scala:
```diff
@@ -111,6 +103,18 @@ class VinylDNS @Inject()(
 
+  // Need this guy for user actions, brings the session username and user account into the Action
+  private val userAction = Action.andThen(new ApiAction(userAccountAccessor.get))
+  private val frontendAction = Action.andThen(new FrontendAction(userAccountAccessor.get))
```
why is the frontendAction needed here?  I see it is only used on `serveCredsFile`.

britneywright replied:
just noting this was answered for anyone else reviewing. userAction doesn't redirect, frontendAction does, and in this case we want to redirect.

pauljamescleary commented on modules/portal/app/controllers/VinylDNS.scala:
```diff
@@ -233,32 +211,24 @@ class VinylDNS @Inject()(
-    }
+  def regenerateCreds(): Action[AnyContent] = userAction.async { implicit request =>
+    Future
+      .fromTry(processRegenerate(request.userName))
```
`processRegenerate` takes a `userName`, but then does an account lookup.  However, we already have the user account on the `request`.  We can pass the `request.account` into the `processRegenerate` instead, and then skip the account lookup there.

mitruly commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,113 @@
+    implicit val executionContext: ExecutionContext)
+    extends ActionFunction[Request, UserRequest] {
+
+  private val cacheHeaders = Seq(
```
These cache headers are the exact same as in the `VinylDNS` singleton and `FrontendAction`. One way to reduce code duplication is to pull these into a shared `trait`. `protected` should make this `cacheHeaders` only accessible from classes/objects which extend it.

rebstar6 replied:
to this point, you actually can share the bulk of this code with something like:

```trait UserAccountAction extends ActionFunction[Request, UserRequest]  {
  
  implicit val executionContext: ExecutionContext
  val userLookup: String => IO[Option[User]]

  protected val cacheHeaders = Seq(
    ("Cache-Control", "no-cache, no-store, must-revalidate"),
    ("Pragma", "no-cache"),
    ("Expires", "0"))

  val notLoggedIn: Result

  def userNotFound(un: String): Result

  val locked: Result

  def invokeBlock[A](request: Request[A], block: UserRequest[A] => Future[Result]): Future[Result] =
  // if the user name is not in session, reject
    request.session.get("username") match {
      case None => Future(notLoggedIn)
      case Some(un) =>
        // user name in session, let's get it from the repo
        userLookup(un).unsafeToFuture().flatMap {
          case None =>
            // Odd case, but let's handle with a different error message
            Future.successful(userNotFound(un))
          case Some(user) if user.lockStatus == LockStatus.Locked => Future.successful(locked)
          case Some(user) => block(new UserRequest(un, user, request))
        }
    }
}
```

rebstar6 replied:
then your api action is just something like 
```
class ApiAction(val userLookup: String => IO[Option[User]])(
    implicit val executionContext: ExecutionContext)
    extends UserAccountAction {

  val notLoggedIn: Result =
    Unauthorized("You are not logged in. Please login to continue.").withHeaders(cacheHeaders: _*)

  def userNotFound(un: String): Result =
    NotFound(s"Unable to find user account for user name '$un'").withHeaders(cacheHeaders: _*)

  val locked: Result = Forbidden(s"Account is locked.").withHeaders(cacheHeaders: _*)
}
```

rebstar6 replied:
only thing im not sure on there is how that implicit EC gets passed around - talk to @pauljamescleary about that

rebstar6 commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,113 @@
+  * If the user is locked out, redirect to login screen
+  * Otherwise, load the account into a custom UserAccountRequest and pass into the action
+  */
+class ApiAction(userLookup: String => IO[Option[User]])(
```
not crazy about the class breakdown all in the same file.

Unless theres a compelling reason not to, id say split this out each into its own file in an actions package

rebstar6 commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,113 @@
+              NotFound(s"Unable to find user account for user name '$un'")
+                .withHeaders(cacheHeaders: _*))
+
+          case Some(user) =>
```
de-nest the if here:

```
        userLookup(un).unsafeToFuture().flatMap {
          case None =>
            // Odd case, but let's handle with a different error message
            Future.successful(
              NotFound(s"Unable to find user account for user name '$un'")
                .withHeaders(cacheHeaders: _*))

          case Some(user) if user.lockStatus == LockStatus.Locked =>
            Future.successful(Forbidden(s"Account is locked.").withHeaders(cacheHeaders: _*))
          case Some(user) =>
            block(new UserRequest(un, user, request))
        }
```

rebstar6 commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,113 @@
+                .flashing(VinylDNS.Alerts.error(s"Unable to find user account for user name '$un'"))
+                .withHeaders(cacheHeaders: _*))
+
+          case Some(user) =>
```
same nested if comment here

rebstar6 commented on modules/portal/app/controllers/UserAccountAction.scala:
```diff
@@ -0,0 +1,113 @@
+    // if the user name is not in session, redirect to the login screen
+    request.session.get("username") match {
+      case None =>
+        Future(
```
the others are Future.successful - any reason for this not to be?

rebstar6 commented on modules/portal/app/controllers/VinylDNS.scala:
```diff
@@ -348,308 +302,253 @@ class VinylDNS @Inject()(
           .withHeaders(cacheHeaders: _*)
       })
+    } else {
+      Future(Forbidden("Request restricted to super users only."))
```
when we are just returning known statuses like this, should it instead be in Future.successful?

pauljamescleary replied:
Probably right.  The danger of Future.successful is if the expression fails, you throw an exception, or possibly return a Future.successful(exception).

If the expression is not known to fail, Future.successful is preferable.

rebstar6 commented on modules/portal/app/actions/ApiAction.scala:
```diff
@@ -0,0 +1,63 @@
+    extends ActionFunction[Request, UserRequest]
+    with CacheHeader {
+
+  def invokeBlock[A](request: Request[A], block: UserRequest[A] => Future[Result]): Future[Result] =
```
you can still move the basic logic here into the trait (CacheHeader, but with this id probably call it VinylDnsAction or something), and just override what it does at each step:

```
trait CacheHeader extends ActionFunction[Request, UserRequest] {
  val cacheHeaders = Seq(
    ("Cache-Control", "no-cache, no-store, must-revalidate"),
    ("Pragma", "no-cache"),
    ("Expires", "0"))
  
  val userLookup: String => IO[Option[User]]
  implicit val executionContext: ExecutionContext
  def notLoggedInResult: Future[Result]
  def cantFindAccountResult(un: String): Future[Result]
  def lockedUserResult: Future[Result]

  def invokeBlock[A](request: Request[A], block: UserRequest[A] => Future[Result]): Future[Result] =
  // if the user name is not in session, reject
    request.session.get("username") match {
      case None => notLoggedInResult

      case Some(un) =>
        // user name in session, let's get it from the repo
        userLookup(un).unsafeToFuture().flatMap {
          // Odd case, but let's handle with a different error message
          case None => cantFindAccountResult(un)

          case Some(user) if user.lockStatus == LockStatus.Locked => lockedUserResult

          case Some(user) =>
            block(new UserRequest(un, user, request))
        }
    }
}
```

```
class ApiAction(val userLookup: String => IO[Option[User]])(
    implicit val executionContext: ExecutionContext)
    extends CacheHeader {

  def notLoggedInResult: Future[Result] = Future.successful(
    Unauthorized("You are not logged in. Please login to continue.")
      .withHeaders(cacheHeaders: _*)
  )

  def cantFindAccountResult(un: String): Future[Result] = Future.successful(
    NotFound(s"Unable to find user account for user name '$un'")
      .withHeaders(cacheHeaders: _*))

  def lockedUserResult: Future[Result] = Future.successful(
    Forbidden(s"Account is locked.")
      .withHeaders(cacheHeaders: _*))
}
```

britneywright replied:
I can make a separate trait. CacheHeader is just for the variable that's used in the controllers and actions. I can make a separate trait for the commonalities in the actions.

rebstar6 commented on modules/portal/app/actions/ApiAction.scala:
```diff
@@ -0,0 +1,63 @@
+    // if the user name is not in session, reject
+    request.session.get("username") match {
+      case None =>
+        Future(
```
make the Futuring consistent - doesnt make sense that this is a Future() but the others are Future.successful.

I believe all the already-determined results can be Future.successful

rebstar6 commented on build.sbt:
```diff
@@ -307,6 +307,7 @@ lazy val portal = (project in file("modules/portal")).enablePlugins(PlayScala, A
     libraryDependencies ++= portalDependencies,
     routesGenerator := InjectedRoutesGenerator,
     coverageMinimum := 75,
+    coverageFailOnMinimum := false,
```
why have a minimum if we dont fail on it?

where are we on coverage now in the portal? id rather lower the minimum if we have to now then just not fail when we hit it

rebstar6 replied:
actually, were above 75 as is, this PR is bringing it down. needs tests it would seem :P

britneywright commented:
I'm pretty sure the reason the code coverage dropped so much for the VinylDNS controller in the portal is due to how we refactored the functions. Since we're using custom actions now we no longer needed the withAuthenticatedUser function, and that wrapped most of the logic in the functions. So if a function wasn't tested (like all of the zone functions) it was only counting as one statement miss, but now there are many more statements in each function. From running the coverage reports locally I see there are 182 statements in master, but in my branch there are now 655. 

I've tested everything related to what my PR adds, even for functions that weren't previously tested at all. I've added 55 tests, that increased the coverage in my branch from 46.59% to 59.61%. Adding more tests seems outside the scope of this PR though. I can branch off this PR for that.

rebstar6 commented:
@britneywright tricky, hmm

We definitely wanna keep coverageFailOnMinimum on, because without it it kinda defeats the point of having any kinda coverage at all. what im inclined to do is, if there are full untested files that you think we should add testing for in a followup, $COVERAGE-OFF$ them, and then write the card to add the tests and turn on coverage for that file.

Open to other suggestions though

rebstar6 commented on modules/portal/app/actions/FrontendAction.scala:
```diff
@@ -0,0 +1,57 @@
+    with CacheHeader {
+
+  def notLoggedInResult: Future[Result] =
+    Future(
```
one last thing - should all of these be `Future` or `Future.successful`?

rebstar6 commented:
other than https://github.com/vinyldns/vinyldns/pull/186/files#r221731723, im good here. please make sure to make followup cards to test and bring back coverage on the skipped items
